### ISSUE TITLE
coredump during multiple nemesis (2 repair and 1 rebuild)

### ISSUE THREAD

**Issue by fgelcer** [scylladb/scylla#5759]
5 db nodes
4 loaders

scylla 3.3 - ami-0e13d9a7078e22943
the scenario is longevity-2tb-7days-1Dis-2NonDis-Nemesis (but it actually runs 4 days, not 7)

we run in this scenario 3 parallel nemesis:
* destroy data and then rebuild
* mgmt repair
* no corrupt repair

```
(CoreDumpEvent Severity.CRITICAL): node=Node longevity-tls-2tb-4d-1dis-2nondis-3-db-node-1dcb436d-3 [34.247.165.138 | 10.0.57.173] (seed: False)
corefile_url=
https://storage.cloud.google.com/upload.scylladb.com/core.scylla.996.4b8bb9ba86194eb4ad0b1391696678ad.6495.1581127202000000/core.scylla.996.4b8bb9ba86194eb4ad0b1391696678ad.6495.1581127202000000.gz
backtrace=           PID: 6495 (scylla)
           UID: 996 (scylla)
           GID: 1001 (scylla)
        Signal: 6 (ABRT)
     Timestamp: Sat 2020-02-08 02:00:02 UTC (2min 52s ago)
  Command Line: /usr/bin/scylla --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 0-11
    Executable: /opt/scylladb/libexec/scylla
 Control Group: /
       Boot ID: 4b8bb9ba86194eb4ad0b1391696678ad
    Machine ID: df877a200226bc47d06f26dae0736ec9
      Hostname: longevity-tls-2tb-4d-1dis-2nondis-3-db-node-1dcb436d-3
      Coredump: /var/lib/systemd/coredump/core.scylla.996.4b8bb9ba86194eb4ad0b1391696678ad.6495.1581127202000000
       Message: Process 6495 (scylla) of user 996 dumped core.
                
                Stack trace of thread 6502:
                #0  0x00007faee0e8f625 raise (libc.so.6)
                #1  0x00007faee0e789ae abort (libc.so.6)
                #2  0x0000000002a5bfbf _ZN7seastar21throw_system_error_onEbPKc (scylla)
                #3  0x00000000029f4412 _ZN7seastar7reactor9run_tasksERNS0_10task_queueE (scylla)
                #4  0x00000000029f4620 _ZN7seastar7reactor14run_some_tasksEv (scylla)
                #5  0x0000000002a34de6 _ZN7seastar7reactor14run_some_tasksEv (scylla)
                #6  0x0000000002a43a3d _ZZN7seastar3smp9configureEN5boost15program_options13variables_mapENS_14reactor_configEENKUlvE1_clEv (scylla)
                #7  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #8  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #9  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6517:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1  0x0000000002de57d6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5ac2 operator() (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6514:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1  0x0000000002de57d6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5ac2 operator() (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6509:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1  0x0000000002de57d6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5ac2 operator() (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6501:
                #0  0x00007faee228aab1 timer_settime@@GLIBC_2.3.3 (librt.so.1)
                #1  0x0000000002a35267 _ZN7seastar8internal18cpu_stall_detector11start_sleepEv (scylla)
                #2  0x0000000002a43a3d _ZZN7seastar3smp9configureEN5boost15program_options13variables_mapENS_14reactor_configEENKUlvE1_clEv (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6516:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1  0x0000000002de57d6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5ac2 operator() (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6520:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1  0x0000000002de57d6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5ac2 operator() (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6513:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1  0x0000000002de57d6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5ac2 operator() (scylla)
                #3  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #5  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6497:
                #0  0x00000000029fc9bf _ZN7seastar17smp_message_queue19flush_request_batchEv (scylla)
                #1  0x0000000002a56196 _ZN7seastar7reactor10smp_pollfn4pollEv (scylla)
                #2  0x00000000029e7a7d _ZN7seastar7reactor9poll_onceEv (scylla)
                #3  0x0000000002a34e19 _ZNKSt8functionIFbvEEclEv (scylla)
                #4  0x0000000002a43a3d _ZZN7seastar3smp9configureEN5boost15program_options13variables_mapENS_14reactor_configEENKUlvE1_clEv (scylla)
                #5  0x00000000029cecae _ZNKSt8functionIFvvEEclEv (scylla)
                #6  0x00007faee18b84e2 start_thread (libpthread.so.0)
                #7  0x00007faee0f54693 __clone (libc.so.6)
                
                Stack trace of thread 6515:
                #0  0x00007faee18c287c read (libpthread.so.0)
                #1

download_instructions=
gsutil cp gs://upload.scylladb.com/core.scylla.996.4b8bb9ba86194eb4ad0b1391696678ad.6495.1581127202000000/core.scylla.996.4b8bb9ba86194eb4ad0b1391696678ad.6495.1581127202000000.gz .
gunzip /var/lib/systemd/coredump/core.scylla.996.4b8bb9ba86194eb4ad0b1391696678ad.6495.1581127202000000.gz
```

**Comment by roydahan** [comment]
@fgelcer please add the decoded backtrace.

Also, we had 2 coredumps in that test, did the have the same backtrace?
What was the scenario (what nemesis) when we add the second coredump?

**Comment by fgelcer** [comment]
> @fgelcer please add the decoded backtrace.
> 
> Also, we had 2 coredumps in that test, did the have the same backtrace?
> What was the scenario (what nemesis) when we add the second coredump?

decoded:
```
2020-02-08 02:00:02.000: (DatabaseLogEvent Severity.ERROR): type=BACKTRACE regex=backtrace line_number=776835 node=Node longevity-tls-2tb-4d-1dis-2nondis-3-db-node-1dcb436d-3 [34.247.165.138 | 10.0.57.173] (seed: False)
2020-02-08T02:00:02+00:00  longevity-tls-2tb-4d-1dis-2nondis-3-db-node-1dcb436d-3 !INFO    | scylla: Backtrace:

void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/fmt/format.h:1158
seastar::backtrace_buffer::append_backtrace() at /usr/include/fmt/format.h:1158
 (inlined by) print_with_backtrace at ./build/release/seastar/./seastar/src/core/reactor.cc:742
seastar::print_with_backtrace(char const*) at /usr/include/fmt/format.h:1158
sigabrt_action at /usr/include/fmt/format.h:1158
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3406
 (inlined by) _FUN at ./build/release/seastar/./seastar/src/core/reactor.cc:3402
?? ??:0
?? ??:0
?? ??:0
seastar::throw_system_error_on(bool, char const*) at /usr/include/fmt/format.h:1158
 (inlined by) seastar::file_desc::sendmsg(msghdr const*, int) at ./build/release/seastar/./seastar/include/seastar/core/posix.hh:257
 (inlined by) seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}::operator()() at ./build/release/seastar/./seastar/src/core/reactor.cc:369
 (inlined by) seastar::apply_helper<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, std::tuple<>&&, std::integer_sequence<unsigned long> >::apply({lambda()#1}&&, std::tuple<>) at ./build/release/seastar/./seastar/include/seastar/core/apply.hh:36
 (inlined by) auto seastar::apply<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}>(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&, std::tuple<>&&) at ./build/release/seastar/./seastar/include/seastar/core/apply.hh:44
 (inlined by) seastar::future<unsigned long> seastar::futurize<seastar::future<unsigned long> >::apply<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}>(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&, std::tuple<>&&) 
at ./build/release/seastar/./seastar/include/seastar/core/future.hh:1645
 (inlined by) seastar::future<unsigned long> seastar::future<>::then_impl<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, seastar::future<unsigned long> >(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&)::{lambda()#1
}::operator()() const::{lambda(seastar::future_state<>&&)#1}::operator()(seastar::future_state) at ./build/release/seastar/./seastar/include/seastar/core/future.hh:1191
 (inlined by) seastar::continuation<seastar::future<unsigned long> seastar::future<>::then_impl<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, seastar::future<unsigned long> >(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambd
a()#1}&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<>&&)#1}>::run_and_dispose() at ./build/release/seastar/./seastar/include/seastar/core/future.hh:506
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/include/fmt/format.h:1158
seastar::reactor::run_some_tasks() at /usr/include/fmt/format.h:1158
seastar::reactor::run_some_tasks() at /usr/include/fmt/format.h:1158
 (inlined by) seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2686
seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}::operator()() const at /usr/include/fmt/format.h:1158
std::function<void ()>::operator()() const at /usr/include/c++/9/bits/std_function.h:690
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:52
?? ??:0
?? ??:0
```

**Comment by fgelcer** [comment]
> @fgelcer please add the decoded backtrace.
> 
> Also, we had 2 coredumps in that test, did the have the same backtrace?
> What was the scenario (what nemesis) when we add the second coredump?

opened a new issue for the other core, as it doesn't have the same stack - #5765

**Comment by slivne** [comment]
@roydahan we have a new flag abort_on_ebadf please add it also the commands you use - it may be a bad file descriptor and lets have this running with it.

**Comment by slivne** [comment]
@bhalevy / @tgrabiec any better idea ?

**Comment by slivne** [comment]
QA are running with the flag

**Comment by slivne** [comment]
@yarongilor / @roydahan if this was reproduced with the flag enabled please update the issue - so benny can continue

**Comment by yarongilor** [comment]
A coredump is reproduced running with the ebadf flag.
test details:
```
ami_id_db_scylla: ami-03cb8ec88d4f2c5fb
'scylla --version' output: 3.3.rc1-0.20200209.0d0c1d43188

18:27:58  < t:2020-02-11 16:27:57,884 f:sct_config.py   l:994  c:sdcm.sct_config      p:INFO  > append_scylla_args: --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc
18:27:58  < t:2020-02-11 16:27:57,884 f:sct_config.py   l:994  c:sdcm.sct_config      p:INFO  >   --abort-on-internal-error 1 --abort-on-ebadf 1
```
coredump details:
```
2020-02-13 01:32:59.000: (CoreDumpEvent Severity.CRITICAL): node=Node longevity-tls-2tb-4d-1dis-2nondis-3-db-node-4f126310-1 [52.30.180.185 | 10.0.64.116] (seed: True)
corefile_url=
https://storage.cloud.google.com/upload.scylladb.com/core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.6064.1581557579000000/core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.
6064.1581557579000000.gz
backtrace=           PID: 6064 (scylla)
           UID: 996 (scylla)
           GID: 1001 (scylla)
        Signal: 6 (ABRT)
     Timestamp: Thu 2020-02-13 01:32:59 UTC (2min 40s ago)
  Command Line: /usr/bin/scylla --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --abort-on-ebadf 1 --log-
to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 0-11
    Executable: /opt/scylladb/libexec/scylla
 Control Group: /
       Boot ID: 7454ba067b074dcc9f8bb647dd0df628
    Machine ID: df877a200226bc47d06f26dae0736ec9
      Hostname: longevity-tls-2tb-4d-1dis-2nondis-3-db-node-4f126310-1
      Coredump: /var/lib/systemd/coredump/core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.6064.1581557579000000
       Message: Process 6064 (scylla) of user 996 dumped core.
                
                Stack trace of thread 6075:
                #0  0x00007fac19427625 raise (libc.so.6)
                #1  0x00007fac194109ae abort (libc.so.6)
                #2  0x0000000002a5bf8f _ZN7seastar21throw_system_error_onEbPKc (scylla)
                #3  0x00000000029f43e2 _ZN7seastar7reactor9run_tasksERNS0_10task_queueE (scylla)
                #4  0x00000000029f45f0 _ZN7seastar7reactor14run_some_tasksEv (scylla)
                #5  0x0000000002a34db6 _ZN7seastar7reactor14run_some_tasksEv (scylla)
                #6  0x0000000002a43a0d _ZZN7seastar3smp9configureEN5boost15program_options13variables_mapENS_14reactor_configEENKUlvE1_clEv (scylla)
                #7  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #8  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #9  0x00007fac194ec693 __clone (libc.so.6)
                
              Stack trace of thread 6087:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6076:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6078:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6086:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6080:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)

                Stack trace of thread 6080:
                #0  0x00007fac19e5a87c read (libpthread.so.0)


                
                Stack trace of thread 6077:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6085:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6082:
                #0  0x00007fac19e5a87c read (libpthread.so.0)
                #1  0x0000000002de57a6 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002de5a92 operator() (scylla)
                #3  0x00000000029cec7e _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fac19e504e2 start_thread (libpthread.so.0)
                #5  0x00007fac194ec693 __clone (libc.so.6)
                
                Stack trace of thread 6074:
                #0  0x0000000000a0b226 _ZN4data12value_writerI21single_fragment_rangeIL12mutable_view0EEEC4ES3_mb (scylla)
                #1  0x0000000000e8b208 _ZN8sstables16make_atomic_cellERK13abstract_typelSt17basic_string_viewIaSt11char_traitsIaEENSt6chrono8durationIlSt5ratioILl1ELl1EEEENS7_10time_pointI8gc_clockSB_EEN7seastar10bool_classIN11atomic_cell21collection_member_tagEEE (scylla)
                #2  0x00000

download_instructions=
gsutil cp gs://upload.scylladb.com/core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.6064.1581557579000000/core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.6064.1581557579000000.gz .
gunzip /var/lib/systemd/coredump/core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.6064.1581557579000000.gz

```

**Comment by bhalevy** [comment]
The `pollable_fd_state::fd` looks corrupt.
```
gdb -c core.scylla.996.7454ba067b074dcc9f8bb647dd0df628.6064.1581557579000000 scylla-package/libexec/scylla
(gdb) bt
#0  0x00007fac19427625 in raise () from /data/5759/scylla-package/libreloc/libc.so.6
#1  0x00007fac194109ae in abort () from /data/5759/scylla-package/libreloc/libc.so.6
#2  0x0000000002a5bf8f in throw_system_error_on (what_arg=<optimized out>, condition=<optimized out>) at ./seastar/include/seastar/core/posix.hh:423
#3  sendmsg (this=<optimized out>, flags=<optimized out>, msg=<optimized out>) at ./seastar/include/seastar/core/posix.hh:257
#4  operator() (__closure=0x60b020df52b0) at ./seastar/src/core/reactor.cc:369
#5  apply (args=<optimized out>, func=<unknown type in /data/5759/scylla-package/libexec/scylla, CU 0x46e45c89, DIE 0x470aa2f8>) at ./seastar/include/seastar/core/apply.hh:36
#6  apply<seastar::pollable_fd_state::write_some(seastar::net::packet&)::<lambda()> > (args=<optimized out>, func=<unknown type in /data/5759/scylla-package/libexec/scylla, CU 0x46e45c89, DIE 0x470aa2c3>) at ./seastar/include/seastar/core/apply.hh:44
#7  apply<seastar::pollable_fd_state::write_some(seastar::net::packet&)::<lambda()> > (args=<unknown type in /data/5759/scylla-package/libexec/scylla, CU 0x46e45c89, DIE 0x470aa281>, func=<unknown type in /data/5759/scylla-package/libexec/scylla, CU 0x46e45c89, DIE 0x470aa28e>)
    at ./seastar/include/seastar/core/future.hh:1645
#8  operator() (state=<unknown type in /data/5759/scylla-package/libexec/scylla, CU 0x46e45c89, DIE 0x470aa1c2>, this=0x60b020df5298) at ./seastar/include/seastar/core/future.hh:1191
#9  seastar::continuation<seastar::future<unsigned long> seastar::future<>::then_impl<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, seastar::future<unsigned long> >(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<>&&)#1}>::run_and_dispose() (this=0x60b020df5280) at ./seastar/include/seastar/core/future.hh:506
#10 0x00000000029f43e2 in seastar::reactor::run_tasks (this=this@entry=0x60b000020000, tq=...) at ./seastar/src/core/reactor.cc:2118
#11 0x00000000029f45f0 in seastar::reactor::run_some_tasks (this=this@entry=0x60b000020000) at ./seastar/src/core/reactor.cc:2541
#12 0x0000000002a34db6 in run_some_tasks (this=0x60b000020000) at ./seastar/src/core/reactor.cc:2686
#13 seastar::reactor::run() () at ./seastar/src/core/reactor.cc:2686
#14 0x0000000002a43a0d in seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}::operator()() const () at ./seastar/include/seastar/core/reactor.hh:744
#15 0x00000000029cec7e in operator() (this=<optimized out>) at /usr/include/c++/9/bits/std_function.h:685
#16 seastar::posix_thread::start_routine (arg=<optimized out>) at ./seastar/src/core/posix.cc:52
#17 0x00007fac19e504e2 in start_thread () from /data/5759/scylla-package/libreloc/libpthread.so.0
#18 0x00007fac194ec693 in clone () from /data/5759/scylla-package/libreloc/libc.so.6
(gdb) fr 4
#4  operator() (__closure=0x60b020df52b0) at ./seastar/src/core/reactor.cc:369
369	        auto r = fd.sendmsg(&mh, MSG_NOSIGNAL);
(gdb) p p
$5 = (seastar::net::packet &) @0x60b050ef8cd0: {static internal_data_size = 112, static default_nr_frags = 4, _impl = {_M_t = {
      _M_t = {<std::_Tuple_impl<0, seastar::net::packet::impl*, std::default_delete<seastar::net::packet::impl> >> = {<std::_Tuple_impl<1, std::default_delete<seastar::net::packet::impl> >> = {<std::_Head_base<1, std::default_delete<seastar::net::packet::impl>, true>> = {<std::default_delete<seastar::net::packet::impl>> = {<No data fields>}, <No data fields>}, <No data fields>}, <std::_Head_base<0, seastar::net::packet::impl*, false>> = {_M_head_impl = 0x60b00b93b7e0}, <No data fields>}, <No data fields>}}}}
(gdb) p mh
$1 = {msg_name = 0x0, msg_namelen = 0, msg_iov = 0x60b00b93b880, msg_iovlen = 3, msg_control = 0x0, msg_controllen = 0, msg_flags = 0}
(gdb) p iov[0]
value has been optimized out
(gdb) p mh.msg_iov[0]
$2 = {iov_base = 0x60b03b4b9c60, iov_len = 90}
(gdb) p mh.msg_iov[1]
$3 = {iov_base = 0x60b009530600, iov_len = 1503}
(gdb) p mh.msg_iov[2]
$4 = {iov_base = 0x60b0005edf50, iov_len = 9}
(gdb) p fd
$6 = {_fd = 1442887936}
(gdb) p/x 1442887936
$10 = 0x5600b900
```
It looks suspiciously like an address, aligned on 256 bytes.

**Comment by bhalevy** [comment]
```
(gdb) scylla apply scylla find 0x5600b900
```
found nothing

**Comment by glommer** [comment]
I've just tackled similar issues in my work with reworking our pollers.
I don't want to throw you into a tangent, but what the problem was for me was that the pollable_fd would be destroyed upon shutdown, leaving stuff alive.

It could be that we're dealing with a similar scenario, in which the pollable_fd_state is already destroyed due to a shutdown, and we're just touching garbage.

**Comment by bhalevy** [comment]
@glommer could very well be.

@scylladb/qa-team: as discussed in the past, we really need to run the nemesis tests also with scylla in debug mode to help find use-after-free and other issues detectable by the sanitizers.

**Comment by glommer** [comment]
@bhalevy debug mode is unusable.

@espindola and I talked in the past about a debug-lite mode, that still can catch *something* but takes less than 2 weeks to boot.

Maybe now it's a good time to revisit that idea. If not, I'd like to insist that we chase this corruption even lacking info. It's hard and it sucks, but better do it now then when it happens in a user site.

**Comment by bhalevy** [comment]
yes, we need to do both.
We have the sanitize mode that's supposed to be the debug-lite version, but we don't even produce relocatable package for it yet.
Regarding the usability of debug mode, it's actually becoming better and the amount of dtests that pass in debug mode keeps improving rapidly and it helps us immensely to find this kind of obscure bugs way before they reach the field.  All in all I think it's definitely worth trying to reproduce in debug mode in parallel to chasing it in release mode (not instead of)

**Comment by bhalevy** [comment]
The file_desc was definitely overwritten:
```
(gdb) p/x *(void**)0x60b00e233a80
$8 = 0x60b05600b900

(gdb) scylla find 0x60b00e233a80
thread 1, small (size <= 80), live (0x60b01529ff50 +56)
thread 1, small (size <= 64), live (0x60b020df5280 +48)

(gdb) p/a (*(void***)0x60b020df5280)[0]
$16 = 0x2a4ab00 <seastar::continuation<seastar::future<>::then_impl<seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}, seastar::future<unsigned long> >(seastar::pollable_fd_state::write_some(seastar::net::packet&)::{lambda()#1}&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<>&&)#1}>::~continuation()>
(gdb) p/a (*(void***)0x60b01529ff50)[0]
$17 = 0x29e8ef0 <seastar::continuation<seastar::future<T>::then_impl(Func&&) [with Func = seastar::pollable_fd_state::write_all(seastar::net::packet&)::<lambda(size_t)>; Result = seastar::future<>; T = {long unsigned int}]::<lambda()>::<lambda(seastar::future_state<long unsigned int>&&)>, long unsigned int>::~continuation(void)>
```

**Comment by eliransin** [comment]
Looks like a networking issue,
According to @roydahan all scenarios ran with TLS encryption server and client sides.
@gleb-cloudius @elcallio can you have a look?

**Comment by espindola** [comment]
> @bhalevy debug mode is unusable.
> 
> @espindola and I talked in the past about a debug-lite mode, that still can catch _something_ but takes less than 2 weeks to boot.

We now have sanitize mode which is debug + opt, but it is still way to slow for unknown reasons.
My first suggestion is to just enable stack guards. They are fairly cheap and I am trying to get them enabled everywhere.
I am also working on a patch series to drop a few calls to need_preempt, which should help debug and sanitize performance as need_preempt always returns true in those modes.

**Comment by elcallio** [comment]
@eliransin does the same run complete w.o. tls active?

**Comment by elcallio** [comment]
I am pretty sure that tls sockets/streams do not prematurely kill underlying fd:s in themselves, though it is possible that calling code not properly keeping a stream alive causes crashes only w. tls on. 
Would still like to know whether this passes w.o. tls. 

Is there _any_ sort of small-scale repro case for this?

**Comment by elcallio** [comment]
Also, is this maybe related to #5765?

**Comment by bhalevy** [comment]
@elcallio I speculate this could be related to `vec_push_wrapper` being called from the gnutls layer (it is registered by `gnutls_transport_set_vec_push_function(*this, &vec_push_wrapper)` from session constructor) after "network reject inter-node communication" (mentioned in #5765)
that causes the stream to be destroyed.

**Comment by slivne** [comment]
@roydahan was this run on 3.2 - is this a regression ?

@roydahan - can we run the same without tls ?

**Comment by slivne** [comment]
@fgelcer we want to try and reproduce outside of SCT what are the exact steps that where done to get the crash ...
- how many nodes
- what is the scylla.yaml
- what operational actions were executed on which node ? and in which order / delay

**Comment by fgelcer** [comment]
> @fgelcer we want to try and reproduce outside of SCT what are the exact steps that where done to get the crash ...
> 
> * how many nodes
> * what is the scylla.yaml
> * what operational actions were executed on which node ? and in which order / delay

sorry for the delayed reply:
* nodes:
** 5 db nodes
** 4 loaders
* scylla.yaml
[scylla.yaml.zip](https://github.com/scylladb/scylla/files/4230835/scylla.yaml.zip)
* actions:
** ManagementRepair nemesis
** CorruptThenRebuild nemesis
** NoCorruptRepair nemesis

**Comment by bhalevy** [comment]
> > @fgelcer we want to try and reproduce outside of SCT what are the exact steps that where done to get the crash ...
> > 
> > * how many nodes
> > * what is the scylla.yaml
> > * what operational actions were executed on which node ? and in which order / delay
> 
> sorry for the delayed reply:
> 
> * nodes:
>   ** 5 db nodes
>   ** 4 loaders
> * scylla.yaml
>   [scylla.yaml.zip](https://github.com/scylladb/scylla/files/4230835/scylla.yaml.zip)
> * actions:
>   ** ManagementRepair nemesis
>   ** CorruptThenRebuild nemesis
>   ** NoCorruptRepair nemesis

@fgelcer those actions don't mean much to us.
Is there a document that describes them in technical terms or code that implements them so was can try to reproduce using dtest?
Also, does it reproduce in SCT?

**Comment by bhalevy** [comment]
@fgelcer in addition to the above, in https://github.com/scylladb/scylla/issues/5765#issue-562383418, that for forked off of this issue, you said:
> we run in this scenario 3 parallel nemesis:
>
> - major compaction
> - network reject inter-node communication
> - stop_wait_start_scylla

Is there a chance that "network reject inter-node communication" was applied here too?
Again, I'm don't know what it does exactly but it seems to be way more related to the symptom than repair.

**Comment by bhalevy** [comment]
@fgelcer can we try to reproduce without tls?
Also, please answer above question.

**Comment by fgelcer** [comment]
> @fgelcer in addition to the above, in [#5765 (comment)](https://github.com/scylladb/scylla/issues/5765#issue-562383418), that for forked off of this issue, you said:
> 
> > we run in this scenario 3 parallel nemesis:
> > 
> > * major compaction
> > * network reject inter-node communication
> > * stop_wait_start_scylla
> 
> Is there a chance that "network reject inter-node communication" was applied here too?
> Again, I'm don't know what it does exactly but it seems to be way more related to the symptom than repair.

it was found on SCT... and i still didn't have the chance to check when happened the inter-node network nemesis.
FYI [here is the code](https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/nemesis.py#L1541) for inter-node network disconnection.

**Comment by slivne** [comment]
Lets try and run this on 3.2 to figure out if its a regression or not.

**Comment by elcallio** [comment]
@bhalevy gnutls should not send or do anything with the wire that is not instigated by caller (us). Data cannot come in without us issuing some read, and nothing will be sent without either using any of the handshake/heartbeat/data carrier functions. There is afaik no spurious calls from within a read call to a send, caused by gnutls state machine. It just decodes things and tells caller.

**Comment by fgelcer** [comment]
there is build ongoing under branch-3.3 (because it was copied from job of branch-3.3), inside staging a job running against latest 3.2 AMI.

**Comment by yarongilor** [comment]
similar reproduction on another 2TB longevity run:

```
2020-02-26 14:50:22.000: (CoreDumpEvent Severity.CRITICAL): node=Node longevity-tls-2tb-4d-1dis-2nondis-3-db-node-835dca26-5 [52.214.95.151 | 10.0.22.68] (seed: False)
corefile_url=
https://storage.cloud.google.com/upload.scylladb.com/core.scylla.996.cacf67514f5443458f5509605a075a41.6726.1582728622000000/core.scylla.996.cacf67514f5443458f5509605a075a41.6726.1582728622000000.gz
backtrace=           PID: 6726 (scylla)
           UID: 996 (scylla)
           GID: 1001 (scylla)
        Signal: 6 (ABRT)
     Timestamp: Wed 2020-02-26 14:50:22 UTC (2min 41s ago)
  Command Line: /usr/bin/scylla --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --abort-on-ebadf 1 --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 0-11
    Executable: /opt/scylladb/libexec/scylla
 Control Group: /
       Boot ID: cacf67514f5443458f5509605a075a41
    Machine ID: df877a200226bc47d06f26dae0736ec9
      Hostname: longevity-tls-2tb-4d-1dis-2nondis-3-db-node-835dca26-5
      Coredump: /var/lib/systemd/coredump/core.scylla.996.cacf67514f5443458f5509605a075a41.6726.1582728622000000
       Message: Process 6726 (scylla) of user 996 dumped core.
                
                Stack trace of thread 6742:
                #0  0x00007fc0ca584625 raise (libc.so.6)
                #1  0x00007fc0ca56d9ae abort (libc.so.6)
                #2  0x00000000029c7622 on_allocation_failure (scylla)
                #3  0x000000000066e1d4 _ZN7seastar6memory16allocate_alignedEmm (scylla)
                #4  0x0000000000ca66cf _ZN7seastar23allocate_aligned_bufferIcEESt10unique_ptrIA_T_NS_12free_deleterEEmm (scylla)
                #5  0x0000000000ca782b _ZN7seastar20shared_ptr_count_forI14zstd_processorEC4IJRKSt8functionIFSt8optionalINS_13basic_sstringIcjLj15ELb1EEEERKS7_EEEEEDpOT_ (scylla)
                #6  0x0000000000ca6bfe _ZNSt17_Function_handlerIFN7seastar10shared_ptrI10compressorEERKSt8functionIFSt8optionalINS0_13basic_sstringIcjLj15ELb1EEEERKS7_EEEPSF_E9_M_invokeERKSt9_Any_dataSE_ (scylla)
                #7  0x0000000000ca5da9 _ZNKSt8functionIFN7seastar10shared_ptrI10compressorEERKS_IFSt8optionalINS0_13basic_sstringIcjLj15ELb1EEEERKS6_EEEEclESD_ (scylla)
                #8  0x0000000000ca4948 _ZN14class_registryIN7seastar10shared_ptrI10compressorEEJRKSt8functionIFSt8optionalINS0_13basic_sstringIcjLj15ELb1EEEERKS7_EEEE6createIJSE_EEES3_SA_DpOT_ (scylla)
                #9  0x0000000000e47be3 operator() (scylla)
                #10 0x0000000000d01542 _ZN8sstables7sstable11data_streamEmmRKN7seastar17io_priority_classE23reader_resource_trackerN7tracing15trace_state_ptrENS1_13lw_shared_ptrINS1_25file_input_stream_historyEEE (scylla)
                #11 0x0000000000e97dc2 _ZN8sstables17data_consume_rowsINS_27data_consume_rows_context_mEEENS_20data_consume_contextIT_EERK6schemaN7seastar13lw_shared_ptrINS_7sstableEEERNS3_8consumerENSA_15disk_read_rangeEm (scylla)
                #12 0x0000000000eaa0b3 _ZNKSt8functionIFN7seastar6futureIJEEEvEEclEv (scylla)
                #13 0x0000000000abe846 _ZN20flat_mutation_reader4implclENSt6chrono10time_pointIN7seastar12lowres_clockENS1_8durationIlSt5ratioILl1ELl1000EEEEEE (scylla)
                #14 0x0000000000c4bd6c _ZN20flat_mutation_readerclENSt6chrono10time_pointIN7seastar12lowres_clockENS0_8durationIlSt5ratioILl1ELl1000EEEEEE (scylla)
                #15 0x0000000000c4c568 operator() (scylla)
                #16 0x0000000000c4d9ab _ZN22mutation_reader_mergerclENSt6chrono10time_pointIN7seastar12lowres_clockENS0_8durationIlSt5ratioILl1ELl1000EEEEEE (scylla)
                #17 0x0000000000c50862 operator() (scylla)
                #18 0x00000000029ff242 _ZN7seastar7reactor9run_tasksERNS0_10task_queueE (scylla)
                #19 0x00000000029ff450 _ZN7seastar7reactor14run_some_tasksEv (scylla)
                #20 0x0000000002a3fc16 _ZN7seastar7reactor14run_some_tasksEv (scylla)
                #21 0x0000000002a4e86d _ZZN7seastar3smp9configureEN5boost15program_options13variables_mapENS_14reactor_configEENKUlvE1_clEv (scylla)
                #22 0x00000000029d9ade _ZNKSt8functionIFvvEEclEv (scylla)
                #23 0x00007fc0cafad4e2 start_thread (libpthread.so.0)
                #24 0x00007fc0ca649693 __clone (libc.so.6)
                
                Stack trace of thread 6749:
                #0  0x00007fc0cafb787c read (libpthread.so.0)
                #1  0x0000000002df0606 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002df08f2 operator() (scylla)
                #3  0x00000000029d9ade _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fc0cafad4e2 start_thread (libpthread.so.0)
                #5  0x00007fc0ca649693 __clone (libc.so.6)
                
                Stack trace of thread 6753:
                #0  0x00007fc0cafb787c read (libpthread.so.0)
                #1  0x0000000002df0606 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002df08f2 operator() (scylla)
                #3  0x00000000029d9ade _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fc0cafad4e2 start_thread (libpthread.so.0)
                #5  0x00007fc0ca649693 __clone (libc.so.6)
                
                Stack trace of thread 6751:
                #0  0x00007fc0cafb787c read (libpthread.so.0)
                #1  0x0000000002df0606 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002df08f2 operator() (scylla)
                #3  0x00000000029d9ade _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fc0cafad4e2 start_thread (libpthread.so.0)
                #5  0x00007fc0ca649693 __clone (libc.so.6)
                
                Stack trace of thread 6748:
                #0  0x00007fc0cafb787c read (libpthread.so.0)
                #1  0x0000000002df0606 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)
                #2  0x0000000002df08f2 operator() (scylla)
                #3  0x00000000029d9ade _ZNKSt8functionIFvvEEclEv (scylla)
                #4  0x00007fc0cafad4e2 start
download_instructions=
gsutil cp gs://upload.scylladb.com/core.scylla.996.cacf67514f5443458f5509605a075a41.6726.1582728622000000/core.scylla.996.cacf67514f5443458f5509605a075a41.6726.1582728622000000.gz .
gunzip /var/lib/systemd/coredump/core.scylla.996.cacf67514f5443458f5509605a075a41.6726.1582728622000000.gz
```

**Comment by fgelcer** [comment]
> there is build ongoing under branch-3.3 (because it was copied from job of branch-3.3), inside staging a job running against latest 3.2 AMI.

Test on 3.2 failed to reproduce the issue. no coredumps in this execution.
There were few repair issues, that will be investigated and if an issue is found, another issue will be raised.

**Comment by slivne** [comment]
currently not a blocker

**Comment by slivne** [comment]
@roydahan / @fgelcer  https://github.com/scylladb/scylla/issues/5759#issuecomment-588740975

@fgelcer - so its clear on 3.2 you ran the exact same SCT version with the nemesis (including networking) and weren't able to reproduce the issue

Can we please run SCT on 3.3 without TLS and verify this doesn't reproduce - it will point the finger toward TLS

**Comment by avikivity** [comment]
Seastar issue: scylladb/seastar#727

**Comment by elcallio** [comment]
This is maybe caused by scylladb/seastar#727

**Comment by avikivity** [comment]
Already backported everywhere.
